lib/strict-non-empty-containers/src/Data/Map/NonEmpty/Strict/Internal.hs

```diff
@@ -0,0 +1,190 @@
+      -- regressions.
+      --
+      -- See the definition of 'NonEmptyMap' for more details of the invariant.
+      --
+      -- To construct a 'NonEmptyMap', use one of the provided constructors,
+      -- all of which are tested to check that they respect the invariant.
+      --
+      NonEmptyMap
+
+    -- * Construction
+    , fromList
+    , fromMap
+    , singleton
+
+    -- * Deconstruction
+    , toList
+    , toMap
+
+    -- * Insertion
+    , insert
+
+    -- * Deletion
+    , delete
+
+    -- * Lookup
+    , lookup
+
+    -- * Combination
+    , unionWith
+
+    -- * Internal functions
+    , invariantHolds
+
+    ) where
+
+import Prelude hiding
+    ( lookup )
+
+import Control.DeepSeq
+    ( NFData )
+import Data.List.NonEmpty
+    ( NonEmpty (..) )
+import Data.Map.Strict
+    ( Map )
+import GHC.Generics
+    ( Generic (..) )
+
+import qualified Data.Foldable as F
+import qualified Data.Map.Strict as Map
+
+-- | A non-empty map from keys of type 'k' to values of type 'v'.
+--
+data NonEmptyMap k v = NonEmptyMap
+    { least
+        :: !(k, v)
+      -- ^ Invariant: this key must be less than than all the keys in 'rest'.
+    , rest
+        :: !(Map.Map k v)
+    }
+    deriving (Eq, Foldable, Functor, Generic, Show, Traversable)
+
+instance (NFData k, NFData v) => NFData (NonEmptyMap k v)
+
+-- | Builds a non-empty map from a list of key-value pairs.
+--
+-- If the list contains more than one value for the same key, the last value
+-- for the key is retained.
+--
+fromList :: Ord k => NonEmpty (k, v) -> NonEmptyMap k v
+fromList (x :| xs) =
+    F.foldl' (\m (k, v) -> insert k v m) (uncurry singleton x) xs
+
+-- | Builds a non-empty map from the given map.
+--
+-- If the given map is empty, this function returns 'Nothing'.
+--
+fromMap :: Map k v -> Maybe (NonEmptyMap k v)
+fromMap = fmap (uncurry NonEmptyMap) . Map.minViewWithKey
+
+-- | Converts a non-empty map to a list of key-value pairs.
+--
+toList :: NonEmptyMap k v -> NonEmpty (k, v)
+toList m = least m :| Map.toList (rest m)
+
+-- | Converts a non-empty map to an ordinary map.
+--
+toMap :: Ord k => NonEmptyMap k v -> Map k v
+toMap m = uncurry Map.insert (least m) (rest m)
+
+-- | Inserts a new key and value in the map.
+--
+-- If the key is already present in the map, the associated value is replaced
+-- with the supplied value.
+--
+insert :: Ord k => k -> v -> NonEmptyMap k v -> NonEmptyMap k v
+insert k v m
+    | k < fst (least m) =
+        NonEmptyMap (k, v) (uncurry Map.insert (least m) (rest m))
+    | k == fst (least m) =
+        NonEmptyMap (k, v) (rest m)
+    | otherwise =
+        m {rest = Map.insert k v (rest m)}
+
+-- | Deletes a key and its value from the map.
+--
+-- When the key is not a member of the map, the original map is returned.
+--
+-- Since this function may reduce the number of elements to zero, it returns
+-- an ordinary map.
+--
+delete :: Ord k => k -> NonEmptyMap k a -> Map k a
```

Comment:
This seemed to me like it would be difficult to use.
I thought that `Maybe (NonEmptyMap)` would be more usable in general.
Checking how it is actually used (one instance) - the result is immediately changed to Maybe.
So I suggest changing the type of this function to match.

Reply:
I did initially consider the type you suggested, but then later on changed it to match the existing (lazy) non-empty map implementations out there.

But I agree that your type makes more sense in the context that we use it, so I'm happy to change it back.

Fixed in: d5a54cd64c53f2f09c8c4703bbbab08a68d037ec.